src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -108,12 +108,15 @@ public NodeStateTable<T> AsCached()
             if (IsCached)
                 return this;
 
-            var nonRemovedCount = _states.Count(static e => !e.IsRemoved);
+            // It's only safe to remove entries that were removed due to their input being removed,
```

Comment:
We should try and clarify this comment a bit, I think. Something like

`If the input to an entry was removed, we also need to remove the entry. However, if the input was present, but the entry didn't produce any values (or removed them all) we need to keep the empty entry as a placeholder, so that on the next generation pass we can retrieve it as the cached value of the input. `

Although I'm also having a hard time succinctly describing it 😆